_source/_posts/2022-03-15-programming-security-and-why-rust.md

```diff
@@ -0,0 +1,157 @@
+tags: [security, rust]
+tweets:
+  - "Why does safe programming matter? Learn what safe programming is and why Rust is a great choice. #rustlang #rust #programming #security"
+  - "Are you still writing C/C++ code? You will be amazed at the amount issues arising from unsafe code. #rustlang #rust #programming #security"
+image: blog/programming-security-and-why-rust/cover.jpg
+type: awareness
+---
+
+As programmers, how many of you have a good understanding of programming safety or secure programming? It's not the same application security or cyber security. I have to confess; I didn't know a lot about these in the early years of my career, especially since I didn't come from a computer science background. But looking back, I think programming security is something every programmer should be aware of and should be taught at a junior level.
+
+What is safe programming, or to be more precise, what does being safe mean for a programming language? or rather what does unsafe mean? Let's set the context first.
+
+{% include toc.md %}
+
+# Programming safety
+
+> Programming Safety = Memory safety + Type safety + Thread safety
+
+Safety can be categorized into three, memory safety, type safety, and thread safety (or four if you count null safety as a separate item)
+
+## Memory safety
+
+In a memory-safe language, when you access a variable or an item in an array, you can be sure that you are indeed accessing what you meant to or are allowed to access. In other words, you will not be reading or writing into the memory of another variable or pointer by mistake, regardless of what you do in your program.
+
+So, why is this a big deal? Doesn't all major programming languages ensure this?
+
+Yes, to varying extents. But some languages are unsafe by default—for example, C and C++. In C or C++, you can access the memory of another variable by mistake, or you can free a pointer twice; that's called double-free error. Sometimes a program continues to use a pointer after it has been freed, and that's called free after use errors or a dangling pointer error. Such behavior is categorized as undefined behavior as they are unpredictable and causes security vulnerability rather than just crashing the program. In this scenario, a crashing program is a good thing as it won't cause a security vulnerability.
+
+> I call it my billion-dollar mistake. It was the invention of the null reference in 1965
+>
+> - Tony Hoare
+
+Then there is also null safety which is kind of related to memory safety. I come from a Java/JavaScript background, and we are so used to the concept of null. Infamous for being the worst invention in programming. Garbage collected languages need a concept of nothing so that a pointer can be freed when unused. But it also leads to issues and pain, like the null pointer exceptions. Technically this falls under memory safety, but most memory-safe languages still let you use null as a value leading to null pointer errors.
```

Comment:
```suggestion
Then there is also null safety, which is related to memory safety. I come from a Java/JavaScript background, and we are so used to the concept of null. Infamous for being the worst invention in programming, garbage collected languages need a concept of nothing so that a pointer can be freed when unused. But it also leads to issues and pain, like the null pointer exceptions. Technically this falls under memory safety, but most memory-safe languages still let you use null as a value leading to null pointer errors.
```